### PR TITLE
[android] Disable the screen wake lock by default

### PR DESCRIPTION
**Pull request by rtsisyk**
Organic Maps drains the battery like crazy with default settings.

**Steps to reproduce:**

1. Install and run the app.
2. Lock your device and put it on a flat surface, with the screen facing up.
3. Pick up the device OR just tap on the screen.
4. The screen will automatically turn on, this is an Android feature.
5. Organic Maps will never allow the screen to go off after that.

Now, imaging this case happens while the phone is in a pocket. The battery may The battery will completely drain before the user even notice that screen is on. I see Organic Maps draining battery and burning screens on my devices regulary.

The problem is caused by two options which are enabled by default - "Show on the lock screen" and "Allow the screen to sleep". The first option adds the app to the lock screen (which is weird on its own). The second option doesn't allow the phone's screen to go off. That is it.

This patch does the following:

- Inverts "Allow the screen to sleep" option to "Keep the screen on".
- Sets the default value of the option to OFF (=allow screen to sleep).
- Applies the option only in FOLLOW & FOLLOW_ROTATE location modes.
- Forces the screen to keep on during navigation no matter of the option.
- Forces the screen to keep on in PENDING_LOCATION no matter of the option.

To summarize, with enabled "Keep the screen on" the screen will be kept
on in the following cases:

- During navigation.
- FOLLOW, FOLLOW_ROTATE, PENDING_LOCATION location modes.

Without the option, the screen will be kept on only during navigation.

### CODE REVIEW

**Comment by vng** [comment]
Also when the app is on a lock screen, any click on an external link (like Place Page or About -> Website, etc) leads to the lock screen again without any action, like the app crashed. This may confuse users.

We don't have a consensus here, so leave your comments, @biodranik described pros and cons above.

**Comment by rtsisyk** [comment]
Nobody knows how many people have removed the app without complaining. This is just survivorship bias. I have Pixel 4a for development that is almost always lies on the table. I return to my home and find Organic Maps burning the screen on that device. It happens ... always. Another Samsung device in the pocket is very regularly getting hot… I take it out of my pocket and see Organic Maps burning the screen... I don't know how convenient both features are, but their combination is unusable. It obviously contradicts with "no battery drain” slogan.

**Comment by RedAuburn** [comment]
i agree with this, it's the first thing i change when reinstalling

**Comment by poeticus** [comment]
An ordinary user's perspective to provide an alternative view: I really like having Organic Maps show on the lock screen. For walking round a city it's very convenient to have the map show simply by pressing the on/off button, and then just as simple to turn it back off. 

I also have 'enable screen to sleep' turned on so it turns off anyway if left (unlike what the original post suggests).

The isues described clearly are not insignificant should anyone experience them, but I think the value of having OM on the lock screen needs to be better appreciated and accepted in the discussion.

**Comment by pastk** [comment]
Another option is to enable by default the "Allow screen to sleep" setting and leave "Show on the lock screen" enabled by default.
Then the battery draining situation won't be possible. I think blocking a screen to sleep is a something a user should opt in and its needed in some specific use cases only (I assume that in navigation mode screen sleep is blocker regardless of the setting, right?).

Still there are concerns of unobvious behavior when interacting with an app on the lock screen. They could be addressed somehow else, e.g. showing a toast upon screen switch on to remind a user that a phone is locked still.

**Comment by poeticus** [comment]
I don't know how possible / easy this would be but just a thought.

If showing a toast to remind a user that the phone is locked is considered a way forward it could be made less intrusive by only displaying it when tapping the buttons at the bottom of the screen. 

A user only needs to be reminded the device is locked if they are likely to be returned to the lock screen. That appears to be the case only with some options found under three of these four buttons. It would appear that use of the search button would never result in this issue so could be excluded from displaying it.

In this way when pressing the on/off button a user could then view and move around the map and perform searches without seeing the toast. A good user experience.

**Comment by pastk** [comment]
> Better still, would it be possible to display a lock screen reminder on screen switch on as suggested, but allow it to be disabled in settings and include a note to this effect in the toast? That way users will become aware of the lock screen display feature but can choose whether or not to continue seeing the reminder.

Sorry, but it looks like something very insignificant to deserve a separate setting.. If we'll add a setting for every minor thing then the number of settings will become unusable very soon.. So its better to focus on behavior that will be convenient w/o any adjustment from the user :)

**Comment by biodranik** [comment]
What about removing the "allow sleep" option and sleeping by default only if the position is not centered?

**Comment by pastk** [comment]
So the sleep will be blocked with a centered/followed position?
But then it won't solve the original battery drain issue.
Also this behavior is not obvious..

**Comment by biodranik** [comment]
The default OM/MapsWithMe behavior for many years before this change was to keep the screen on only if position is centered (nav mode centers it automatically after 20 or 30 seconds). And in cases when device lock breaks active download.

Are there any other cases when it is really important to keep screen on and don't alow autosleep/autolock?

You can check the git log/PR history behind introducing the "sleep" option.

**Comment by biodranik** [comment]
Updated the list of pros and cons above.

**Comment by rtsisyk** [comment]
Is there any disagreement that having both 'Show on the lock screen' and 'Allow the screen to sleep' enabled simultaneously is a BUG? I frequently discover my devices running out of battery due to this unfortunate combination of default values. This occurs because I reinstall a debug/beta versions quite frequent with resetting all settings to the default values. My test devices are not unique in any way – they are standard Pixel 4a running CalyxOS/Android 13 and Samsung devices running Samsung/Android 12. In my opinion, it is a BUG.

**Comment by rtsisyk** [comment]
> The default OM/MapsWithMe behavior for many years before this change was to keep the screen on only if position is centered (nav mode centers it automatically after 20 or 30 seconds). And in cases when device lock breaks active download.

Ah, I started to understand. One silly question. Are we talking about the same app? Maybe "OM" means something different here, but in Organic Maps on Android (2023.09.24-4 compiled today) `FLAG_KEEP_SCREEN_ON` flag is on the start of the main (MwmActivity). Proof:  https://github.com/organicmaps/organicmaps/blob/634a3182019157cf9ea556535d12c5e2cbe858c4/android/app/src/main/java/app/organicmaps/MwmActivity.java#L1058-L1059

I don't see any difference in the behavior no matter if the position is centered on the screen or not. Neither in the code nor practically in the app. This is why I started to have doubts that we were talking about the same app. The bug addressed by this PR pretty easy to reproduce.  Please watch the video with the RIGHT application. Please make sure that your app looks the same when you test it. 

https://github.com/organicmaps/organicmaps/assets/1799054/7a1f16e6-e787-4aa4-9cf3-d6ee927b6ace

The screen will stay on until the phone fully discharges. All settings are default. The phone was refreshed recently, no special settings. I consider it as a bug. Is there any disagreement?

**Comment by rtsisyk** [comment]
This reported issue about battery drains issue reproduces 100% on any contemporary Android device that support unlocking the screen on touch or on "take the device in your hand" gesture.  While there's a room for continued conversation about enhancing the malfunctioning "show on the lock screen" feature, it's crucial to fix the battery drain issue right now.

**Comment by biodranik** [comment]
There has been only one user complaint in almost two years about this feature. This one.

**Comment by rtsisyk** [comment]
> There has been only one user complaint in almost two years about this feature. This one.

Do you have the number how many users has removed the app because of that?

**Comment by rtsisyk** [comment]
> There has been only one user complaint in almost two years about this feature. This one.

Do you accept the fact that this is a bug?

**Comment by biodranik** [comment]
As I mentioned before, nobody complained about this behavior before. I assume that either people use the gesture to awake OM on the lock screen when it's needed, or their phones don't behave like this. If OM is left on the lock screen, it is often means that you are using OM for some reason. Then it implies that you need/want/expect to see OM again after unlocking the phone, to use it for your use-case. Breaking this default convenient behavior is bad IMO, because most users (especially new ones) won't discover that this cool feature is available in settings and can be enabled (after you change default to disabled).

The opposite situation is not as bad: a few users who may hit issue like you, can disable a setting in OM, or their system gesture-awake setting and continue enjoying using OM. That's why I'm open to any other solution except worsening the very convenient, simple-to-use, OM-style UX default.

**Comment by rtsisyk** [comment]
> As I mentioned before, nobody complained about this behavior before. I assume that either people use the gesture to awake OM on the lock screen when it's needed, or their phones don't behave like this. If OM is left on the lock screen, it is often means that you are using OM for some reason. Then it implies that you need/want/expect to see OM again after unlocking the phone, to use it for your use-case. Breaking this default convenient behavior is bad IMO, because most users (especially new ones) won't discover that this cool feature is available in settings and can be enabled (after you change default to disabled).

Have you seen the video? Do you accept the evidence shown by the video that the phone screen never goes off after being woken up accidentally by a gesture because of the Organic Maps?

**Review by biodranik — CHANGES_REQUESTED** [review]
The fix for this issue should not break the existing convenient UX for many users. Let's focus on other options to resolve it.

**Comment by rtsisyk** [comment]
> The fix for this issue should not break the existing convenient UX for many users. Let's focus on other options to resolve it.

How does this PR break UX for many users? Please elaborate. The option is still available and can be enabled if needed.

**Comment by rtsisyk** [comment]
Updated after discussion to disable the screen wake lock by default. 

> This patch does the following:
>
> -   Inverts "Allow the screen to sleep" option to "Keep the screen on".
> -    Sets the default value of the option to OFF (=allow screen to sleep).
> -    Applies the option only in FOLLOW & FOLLOW_ROTATE location modes.
> -    Forces the screen to keep on during navigation no matter of the option.
> -    Forces the screen to keep on in PENDING_LOCATION no matter of the option.
>
> To summarize, with enabled "Keep the screen on" the screen will be kept
on in the following cases:
>
> -    During navigation.
> -    FOLLOW, FOLLOW_ROTATE, PENDING_LOCATION location modes.
>
> Without the option, the screen will be kept on only during navigation.
>

File: android/app/src/main/java/app/organicmaps/MwmActivity.java [hunk 1]
```diff
@@ -1700,6 +1703,17 @@ public void onMyPositionModeChanged(int newMode)
     if (controller.isPlanning())
       showAddStartOrFinishFrame(controller, true);
     LocationHelper.from(this).restartWithNewMode();
+    switch (newMode) {
+    case PENDING_POSITION:
+      Utils.keepScreenOn(true, getWindow());
+      break;
```
Comment (vng) [inline android/app/src/main/java/app/organicmaps/MwmActivity.java:1709]: So, is it possible if a position is not available, we will get an infinite screen here with "Keep searching" alert?

File: android/app/src/main/java/app/organicmaps/DownloadResourcesLegacyActivity.java [hunk 2]
```diff
@@ -231,7 +232,7 @@ protected void onSafeDestroy()
     super.onSafeDestroy();
     mApiRequest.unregister();
     mApiRequest = null;
-    Utils.keepScreenOn(false, getWindow());
+    Utils.keepScreenOn(Config.isKeepScreenOnEnabled(), getWindow());
```
[the remark this answers is not in the document]
Comment (rtsisyk) [reply android/app/src/main/java/app/organicmaps/DownloadResourcesLegacyActivity.java:235]: This is a good question. I think that the intention was to keep screen on because the downloader doesn't work in the background. I also thought about removing this hack...

File: android/app/src/main/java/app/organicmaps/MwmActivity.java (hunk 1 above)
Comment (rtsisyk) [reply android/app/src/main/java/app/organicmaps/MwmActivity.java:1709]: It is... Probably it wasn't a good idea. Instead, we start a foreground service with a notification to show that the location search is in progress. I added an alternative solution for PENDING_POSITION (https://github.com/organicmaps/organicmaps/pull/6436) and probably can add a foreground service & notification later. This is out of scope of this PR and I am removing keepScreenOn() for PENDING_POSITION from the code.

File: android/app/src/main/java/app/organicmaps/DownloadResourcesLegacyActivity.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (rtsisyk) [reply android/app/src/main/java/app/organicmaps/DownloadResourcesLegacyActivity.java:235]: It looks like that this hack is still needed, because otherwise the screen goes off during DownloadResourcesLegacyActivity (the world.mwm downloader) and aborts downloading... Yes, a better solution is needed, but let's address it separately. This PR doesn't change the behavior there.

File: android/app/src/main/java/app/organicmaps/MwmActivity.java (hunk 1 above)
Comment (rtsisyk) [reply android/app/src/main/java/app/organicmaps/MwmActivity.java]: Removed the wake lock for PENDING_POSITION mode.

**Review by vng — APPROVED** [review]
Android checks show something wrong with string ids.

File: android/app/src/main/java/app/organicmaps/DownloadResourcesLegacyActivity.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (biodranik) [reply android/app/src/main/java/app/organicmaps/DownloadResourcesLegacyActivity.java:235]: There are many ways to keep a specific window/view on without wakelocks, why they can't be applied to the world mwm downloader? https://stackoverflow.com/a/18487237

[the remark this answers is not in the document]
Comment (rtsisyk) [reply android/app/src/main/java/app/organicmaps/DownloadResourcesLegacyActivity.java:235]: Downloader should use a foreground service. But it is a different issue.

**Comment by rtsisyk** [comment]
> Android checks show something wrong with string ids.

It just required re-generation of strings.

File: data/strings/strings.txt [hunk 3]
```diff
@@ -26156,90 +26156,92 @@
...
+    ja = 画面を表示したままにする
+    ko = 화면 켜짐 유지
+    mr = स्क्रीन चालू ठेवा
+    nb = Behold skjermen på
+    nl = Scherm ingeschakeld houden
+    pl = Pozostaw ekran włączony
+    pt = Mantém o ecrã ligado
+    pt-BR = Manter a tela ligada
+    ro = Păstrați ecranul pornit
+    ru = Держи экран включенным
```
Comment (biodranik) [inline data/strings/strings.txt:26191]: ru = Не выключать экран

File: data/strings/strings.txt [hunk 4]
```diff
@@ -26156,90 +26156,92 @@
...
-    sv = Låt skärmen sova
-    th = อนุญาตให้หน้าจอเข้าสู่โหมดสลีป
-    tr = Ekranın kapanmasına izin ver
-    uk = Дозволити екрану вимкнутись
-    vi = Cho phép màn hình ở chế độ ngủ
-    zh-Hans = 允许屏幕进入休眠状态
-    zh-Hant = 允许螢幕進入休眠狀態
-
-  [enable_screen_sleep_description]
+    en = Keep the screen on
```
Comment (biodranik) [inline data/strings/strings.txt:26161]: @organicmaps/translations Please review these translations.

Comment (biodranik) [inline data/strings/strings.txt:26161]: @rtsisyk please make a follow-up fix.

tag = android
comment = Setting in preferences to completely disable screen dimming and auto device lock
